File: src/micapolos/zexy/Color.java
package micapolos.zexy;

import micapolos.tata8.Canvas;

import static micapolos.Leo.*;
import static micapolos.zexy.Number.*;

public final class Color extends Component implements Drawable {
  public final Number red;
  public final Number green;
  public final Number blue;
  public final Number alpha;

  Color(Number red, Number green, Number blue, Number alpha) {
    this.red = red;
    this.green = green;
    this.blue = blue;
    this.alpha = alpha;
  }

  @Override
  void addRunners() {
    red.addRunnersOnce();
    green.addRunnersOnce();
    blue.addRunnersOnce();
    alpha.addRunnersOnce();
  }

  public static Color rgb(double red, double green, double blue) {
    return rgba(red, green, blue, 1);
  }

  public static Color rgba(double red, double green, double blue, double alpha) {
    return rgba(Number.number(red), Number.number(green), Number.number(blue), Number.number(alpha));
  }

  public static Color rgb(Number red, Number green, Number blue) {
    return rgba(red, green, blue, Number.number(1));
  }

  public static Color rgba(Number red, Number green, Number blue, Number alpha) {
    return new Color(red, green, blue, alpha);
  }

  public static Color variable() {
    return new Color(Number.newNumber(), Number.newNumber(), Number.newNumber(), Number.newNumber());
  }

  public Action set(Color color) {
    return Action.sequence(
      red.set(color.red),
      green.set(color.green),
      blue.set(color.blue),
      alpha.set(color.alpha));
  }

  public micapolos.tata8.Color get() {
    return micapolos.tata8.Color.rgba(
      (float) red.get(),
      (float) green.get(),
      (float) blue.get(),
      (float) alpha.get());
  }

  @Override
  public String toString() {
    return leo("color",
      leo("red", red),
      leo("green", green),
      leo("blue", blue),
      leo("alpha", alpha));
  }

  @Override
  public void drawOn(Canvas canvas) {
    canvas.fillRect(128, 96, 64, 64, tata8());
  }

  micapolos.tata8.Color tata8() {
    return micapolos.tata8.Color.rgba((float) red.get(), (float) green.get(), (float) blue.get(), (float) alpha.get());
  }

  static void main() {
    var brightness = Key.Z.isPressed.select(seconds.fraction(), number(1));
    Color.rgb(
        Key.LEFT.isPressed.toNumber().times(brightness),
        Key.DOWN.isPressed.toNumber().times(brightness),
        Key.RIGHT.isPressed.toNumber().times(brightness))
      .show();
  }
}

File: src/micapolos/zexy/Color.kt
package micapolos.zexy

import micapolos.zexy.model.Color as ModelColor

class Color(model: ModelColor): Value<Color>(model)

internal val Value<Color>.modelColor get() = model as ModelColor.Rgba

val color = color(1.0, 1.0, 1.0, 1.0)

fun color(red: Double, green: Double, blue: Double, alpha: Double) =
  color(red.value, green.value, blue.value, alpha.value)

fun color(red: Value<Number>, green: Value<Number>, blue: Value<Number>, alpha: Value<Number>) =
  Color(
    ModelColor.Rgba(
      red.modelNumber,
      green.modelNumber,
      blue.modelNumber,
      alpha.modelNumber))

val Value<Color>.transparent get() = color(0.0, 0.0, 0.0, 0.0)
val Value<Color>.white get() = color(1.0, 1.0, 1.0, 1.0)
val Value<Color>.red get() = color(1.0, 0.0, 0.0, 1.0)
val Value<Color>.green get() = color(0.0, 1.0, 0.0, 1.0)
val Value<Color>.blue get() = color(1.0, 1.0, 1.0, 1.0)
val Value<Color>.yellow get() = color(1.0, 1.0, 0.0, 1.0)
val Value<Color>.black get() = color(0.0, 0.0, 0.0, 1.0)

fun Value<Color>.withRed(red: Double) = withRed(red.value)
fun Value<Color>.withRed(red: Value<Number>) =
  Color(
    ModelColor.Rgba(
      red.modelNumber,
      modelColor.green,
      modelColor.blue,
      modelColor.alpha))

fun Value<Color>.withGreen(green: Double) = withGreen(green.value)
fun Value<Color>.withGreen(green: Value<Number>) =
  Color(
    ModelColor.Rgba(
      modelColor.red,
      green.modelNumber,
      modelColor.blue,
      modelColor.alpha))

fun Value<Color>.withBlue(blue: Double) = withBlue(blue.value)
fun Value<Color>.withBlue(blue: Value<Number>) =
  Color(
    ModelColor.Rgba(
      modelColor.red,
      modelColor.green,
      blue.modelNumber,
      modelColor.alpha))

fun Value<Color>.withAlpha(alpha: Double) = withAlpha(alpha.value)
fun Value<Color>.withAlpha(alpha: Value<Number>) =
  Color(
    ModelColor.Rgba(
      modelColor.red,
      modelColor.green,
      modelColor.blue,
      alpha.modelNumber))
